### ISSUE TITLE
Actor isolation is broken by conforming to an async protocol

### ISSUE THREAD

**Issue by alexd6631** [apple/swift#58517]
**Describe the bug**

It is possible to break the actor abstraction (one synchronous section of the actor executed at a time) by making it conform to a protocol with async methods.

This bug has been discussed here, and I was advised to report it
https://forums.swift.org/t/actor-isolation-is-broken-by-protocol-conformance/57040

**To Reproduce**

In the example code below, an actor implements two variant of a protocol, one async and one inheriting Actor protocol.

When testing a race between two tasks calling an actor method (simulating an expensive CPU work without suspension point), we can see that the TestableActor  will honor the isolation of the actor, whereas the async one will allow tasks to interleave and break actor invariant.

```swift
protocol TestableActor: Actor {
    func test()
}

protocol TestableAsync {
    func test() async
}

actor TestActor : TestableActor, TestableAsync {
    func test() {
        print("Before \(Thread.current)")
        Thread.sleep(forTimeInterval: 1)
        print("After \(Thread.current)")
    }
}


func test(actor: TestableActor) {
    Task {
        print("TestableActor 1 \(Thread.current)")
        await actor.test()
    }

    Task {
        print("TestableActor 2 \(Thread.current)")
        await actor.test()
    }
}

func testAsync(actor: TestableAsync) {
    Task{
        print("TestableAsync 1 \(Thread.current)")
        await actor.test()
    }

    Task {
        print("TestableAsync 2 \(Thread.current)")
        await actor.test()
    }
}

let actor = TestActor()

test(actor: actor)
Thread.sleep(forTimeInterval: 3)

testAsync(actor: actor)
Thread.sleep(forTimeInterval: 3)
```

Output : 

```
TestableActor 1 <NSThread: 0x1011cbba0>{number = 2, name = (null)}
Before <NSThread: 0x1011cbba0>{number = 2, name = (null)}
TestableActor 2 <NSThread: 0x1010044a0>{number = 3, name = (null)}
After <NSThread: 0x1011cbba0>{number = 2, name = (null)}
Before <NSThread: 0x1011cbba0>{number = 2, name = (null)}
After <NSThread: 0x1011cbba0>{number = 2, name = (null)}
TestableAsync 2 <NSThread: 0x1010044a0>{number = 3, name = (null)}
Before <NSThread: 0x1010044a0>{number = 3, name = (null)}
TestableAsync 1 <NSThread: 0x1011cbba0>{number = 2, name = (null)}
Before <NSThread: 0x1011cbba0>{number = 2, name = (null)} <--- <!> Interleave happens here
After <NSThread: 0x1011cbba0>{number = 2, name = (null)}
After <NSThread: 0x1010044a0>{number = 3, name = (null)}
```

Forcing inheriting the Actor protocol works, but it is a leaking abstraction, and it is not possible to abstract over two implementation using or not an actor as an implementation detail.

As discussed in the forum thread, I know it is not a good practice to use Thread.sleep inside the actor. It is here to simulate a long running synchronous computation like a CPU-intensive task and ease of reproduce.
If you swap the Thread.sleep by a non trivial computation (several ms) you will still observe the problem.

**Expected behavior**

I would expect the protocol witness of the async protocol to preserve the underlying actor isolation without additional information. It should never be possible to break the actor isolation from the outside.

As suggested in the forum thread, we could also consider that Actor is not really implementing TestableAsync, so the compiler would consider it an error, forcing us to add async on Actor test() body, which works in that case.

**Screenshots**

**Environment (please complete the following information):**
 - OS: macOS 12.3.1
 - Xcode 13.3.1

Create a simple macOS CLI and paste the sample code here.

**Additional context**
The problem was found by a race on an iOS application, then I created this minimal example for reproducing.

**Comment by hooman** [comment]
Pinging @ktoso: Can you please take a look and tell us what you think?

**Comment by DougGregor** [comment]
Looking at the generated SIL for

```swift
protocol TestableAsync {
  func test() async
}

actor TestActor : TestableAsync {
  func test() { }
}
```
we are *clearly* missing an actor hop in the witness.

**Comment by DougGregor** [comment]
Also tracking this via rdar://92881539

**Comment by alexd6631** [comment]
Hello,

Any news regarding a fix for this issue in the next Swift 5.7 release ?

Best regards
